src/tokenizer.js

```diff
@@ -0,0 +1,792 @@
+
+export class Token {}
+
+// The output of tokenization step is a stream of zero or more of the following tokens: <ident-token>, <function-token>,
+// <at-keyword-token>, <hash-token>, <string-token>, <bad-string-token>, <url-token>, <bad-url-token>, <delim-token>,
+// <number-token>, <percentage-token>, <dimension-token>, <whitespace-token>, <CDO-token>, <CDC-token>, <colon-token>,
+// <semicolon-token>, <comma-token>, <[-token>, <]-token>, <(-token>, <)-token>, <{-token>, and <}-token>.
+export class IdentToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class FunctionToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class AtKeywordToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class HashToken extends Token {
+  type;
+  value;
+  constructor(value, type = 'unrestricted') {
+    super();
+    this.value = value;
+    this.type = type;
+  }
+}
+
+export class StringToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class BadStringToken extends Token {}
+
+export class UrlToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class BadUrlToken extends Token {}
+
+export class DelimToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class NumberToken extends Token {
+  value;
+  type;
+  constructor(value, type = "integer") {
+    super();
+    this.value = value;
+    this.type = type;
+  }
+}
+
+export class PercentageToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class DimensionToken extends Token {
+  value;
+  type;
+  unit;
+  constructor(value, type, unit) {
+    super();
+    this.value = value;
+    this.type = type;
+    this.unit = unit;
+  }
+}
+
+export class WhitespaceToken extends Token {}
+
+export class CDOToken extends Token {}
+
+export class CDCToken extends Token {}
+
+export class ColonToken extends Token {}
+
+export class SemicolonToken extends Token {}
+
+export class CommaToken extends Token {}
+
+export class LeftSquareBracketToken extends Token {}
+
+export class RightSquareBracketToken extends Token {}
+
+export class LeftParenthesisToken extends Token {}
+
+export class RightParenthesisToken extends Token {}
+
+export class LeftCurlyBracketToken extends Token {}
+
+export class RightCurlyBracketToken extends Token {}
+
+class InputStream {
+  codePoints;
+  index = 0;
+  returned = [];
+  constructor(input) {
+    this.codePoints = Array.from(input).map(char => char.codePointAt(0));
```

Comment:
I'm not sure how often this is called in the polyfill, but I suspect it would be more efficient to keep a reference to the original input and do the mapping to codepoints just in time as accessed by consume or peek.

Reply:
It shouldn't be called outside the initial parsing, as we store the resulting CSSNumericValues, but that's no excuse 😅. I switched to iterating through the input string directly in https://github.com/flackr/scroll-timeline/pull/206/commits/6ddd897a7b1a9fc62883af224a569d1b1df7b502.

Reply:
Looks great!